style context added to roc plot

yafimvo opened the pull request:
## Describe your changes
1. Style context added to control plots styles
2. `apply_theme` decorator added to plot functions
3. Base images updated
4. Modified existing plot hardcoded style

## Issue ticket number and link
Closes #243  

## How does it work?
Context style controls the visual style of our matplotlib plots in one place using the `rcParams`.
Currently, the default style is an "L" shaped chart (only left and bottom borders)  and a monochromatic coloring (tints of one color) based on the material UI color palette. We can modify both with the`cmap_style` and `ax_style` parameters.
* `cmap_style` : ["monochromatic", "gradient"]
* `ax_style` : ["no_frame", "frame"]

For example:
Let's style `_silhouette_analysis_one_model` which generates a silhouette plot for one value of n_cluster.

the default style is:

```python
@apply_theme()
def _silhouette_analysis_one_model(..., ax =None):
    # plot implementation
```
result:
![image](https://user-images.githubusercontent.com/103026689/220566459-02136d05-14b6-4b54-8dd4-3193fa980c18.png)

the default style is:
```python
@apply_theme(cmap_style="gradient")
def _silhouette_analysis_one_model(..., ax =None):
    # plot implementation
```
result:
![image](https://user-images.githubusercontent.com/103026689/220566356-aac950fa-2bfa-4be8-a360-65b8eb2cd6e2.png)


We can override theme by adding custom logic within the method, for example: adding a grid and right border.
```python
@apply_theme(cmap_style="gradient")
def _silhouette_analysis_one_model(..., ax =None):
    # plot implementation
    ax.spines['right'].set_visible(True)
    ax.grid(True)
```
![image](https://user-images.githubusercontent.com/103026689/220567346-523a3906-55ea-40b6-8914-772e92f9205d.png)

## Checklist before requesting a review

- [X] I have performed a self-review of my code
- [X] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [X] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)



<!-- readthedocs-preview sklearn-evaluation start -->
----
:books: Documentation preview :books:: https://sklearn-evaluation--279.org.readthedocs.build/en/279/

<!-- readthedocs-preview sklearn-evaluation end -->

edublancas commented:
looks like this is still wip so I marked it as draft

yafimvo commented:
@edublancas I added a decorator and context manager to style the plots and applied them to the ROC curve.

Let me know if the style is OK with you so I can update the tests and modify more plots.

[Here is an example of the ROC curve](https://sklearn-evaluation--279.org.readthedocs.build/en/279/api/plot.html#roc)

edublancas commented:
looks good!

this can be easily applied to many line plots; but there are others that might need tweaks. for example, confusion matrix. how are you thinking of that one?

yafimvo commented:
> this can be easily applied to many line plots; but there are others that might need tweaks. for example, confusion matrix. how are you thinking of that one?

I added some logic to the decorator and the context so we can control the applied theme, whether it's colors, ax settings, or any other specific configuration we want.

We can apply it on cm as follows:

```python
@apply_theme(ax_style=False)
def plot(self, ax=None, **theme):
    # custom plot creation logic
```

Since we're passing the `cmap` parameter to `_plot_cm` and calculating its value sometimes in the `__init__` method and sometimes in the `plot` method, and sometimes in the `_matrix.add`; I added this `**theme` parameter so we can easily extract the `cmap` from the "decorated" function without adding an extra logic and breaking current stuff.

@edublancas what do you think?

edublancas commented:
> ; I added this **theme parameter so we can easily extract the cmap from the "decorated" function without adding an extra logic and breaking current stuff.

I'm not following how this `**theme` works.

yafimvo commented:
> I'm not following how this `**theme` works.

It allows us to pass the `cmap` parameter via `theme` and use it in [`def _plot_cm`](https://github.com/ploomber/sklearn-evaluation/blob/6aadc171dc9f6ffa11c1c9fd62e96848417ed9a0/src/sklearn_evaluation/plot/classification.py#L288) without breaking the current inner API

```python
im = ax.imshow(cm, interpolation="nearest", cmap=cmap)
ax.figure.colorbar(im)
```

I can remove this `theme` parameter and apply cmap in the context manager, but I will have to change these functions:

* [def _plot_cm](https://github.com/ploomber/sklearn-evaluation/blob/6aadc171dc9f6ffa11c1c9fd62e96848417ed9a0/src/sklearn_evaluation/plot/classification.py#L284)

* [sklearn_evaluation.plot._matrix.add](https://github.com/ploomber/sklearn-evaluation/blob/6aadc171dc9f6ffa11c1c9fd62e96848417ed9a0/src/sklearn_evaluation/plot/_matrix.py#L36)

and some more places that use `default_heatmap`

edublancas commented:
I think let's remove `**theme` for now. I think we can later offer an API to customize the plots but adding generic keyword arguments like `**theme` will be confusing for users.

edublancas commented on src/sklearn_evaluation/plot/feature_ranking.py:
```diff
@@ -196,7 +198,11 @@ def feature_ranks_custom_algorithm(self, ranks):
         self._draw()
         return self.ax
 
+    def _set_default_colors(self, cmap):
```
why is this needed? looks like it'll edit matplotlib's globals

yafimvo replied:
Good catch! removed it.

edublancas commented on src/sklearn_evaluation/plot/feature_ranking.py:
```diff
@@ -394,7 +404,8 @@ def _draw(self):
         # Draw the heatmap
         data = np.ma.masked_where(mask, self.ranks_)
-        mesh = self.ax.pcolormesh(data, cmap=self.colormap, vmin=-1, vmax=1)
+        # mesh = self.ax.pcolormesh(data, cmap=self.colormap, vmin=-1, vmax=1)
```
can we remove this commented line?

edublancas commented on src/sklearn_evaluation/plot/grid_search.py:
```diff
@@ -202,12 +201,17 @@ def _grid_search_single(grid_scores, change, subset, kind, ax, sort, params):
     ax.set_ylabel("Mean score")
     ax.set_xlabel(change)
-    ax.legend(loc="best")
+    ax.legend()
```
Im guessing this was removed since best is the default value?

yafimvo replied:
yes

edublancas commented on src/sklearn_evaluation/plot/matplotlib/bar.py:
```diff
@@ -41,9 +44,9 @@ def horizontal(values, labels=None, sort=True, error=None, ax=None):
+        ax.barh(y_pos, values, edgecolor="#fff")
     else:
-        ax.barh(y_pos, values, xerr=error)
+        ax.barh(y_pos, values, xerr=error, edgecolor="#fff")
```
is there an alternative API that would set this from the `apply_theme` decorator? or do you think it's better to set it here

yafimvo replied:
solved with this:
```
plt.rcParams['patch.edgecolor'] = '#fff'
plt.rcParams['patch.force_edgecolor'] = True
```

edublancas commented on src/sklearn_evaluation/plot/pca.py:
```diff
@@ -114,7 +115,8 @@ def _plot_generic(n_components, principal_components, ax):
             ax[ax_ind].scatter(
-                principal_components[:, pc1], principal_components[:, pc2], s=50
+                principal_components[:, pc1], principal_components[:, pc2],
+                alpha=.6, edgecolor="face", s=50
```
same as my previous comment. is it better to set the edgecolor property inline rather than at the apply_theme level?

yafimvo replied:
According to [matplotlib api](https://matplotlib.org/stable/tutorials/introductory/customizing.html#the-default-matplotlibrc-file) docs it's possible using these attributes:

```
#scatter.marker: o         # The default marker type for scatter plots.
#scatter.edgecolors: face  # The default edge colors for scatter plots.
```

I tried
```
plt.rcParams['scatter.edgecolors'] = 'face'
```

But for some reason, it doesn't work. btw, changing `scatter.marker` works.

edublancas replied:
alright, if it doesn't work, let's harcode it then

edublancas commented on src/sklearn_evaluation/plot/precision_recall.py:
```diff
@@ -19,7 +20,7 @@
     ax.set_title(name)
     ax.set_xlim([0.0, 1.0])
-    ax.set_ylim([0.0, 1.05])
+    ax.set_ylim([0.0, 1.0])
```
any reason to change this?

yafimvo replied:
This adds a small margin at the top. removed this since we don't have an upper border.

Let me know which one is better

![image](https://user-images.githubusercontent.com/103026689/220856203-e58656bc-7ea2-4e46-a4fb-62d5672b93ae.png)

![image](https://user-images.githubusercontent.com/103026689/220856115-8e9b5342-cee9-46b1-a175-3d1715f27848.png)

edublancas replied:
the top one looks better

edublancas reviewed (changes requested):
looks awesome! added a few minor comments.

also, the grid in many plots is gone (e.g., elbow curve). let's bring it back since it helps a lot with visualization. whenever there is a numeric scale in the X or Y axis, having a grid helps users see where the bar/line/dot is

edublancas commented:
also, please briefly document how to use this in the developer's guide: https://sklearn-evaluation.ploomber.io/en/latest/reference/developer.html

tonykploomber commented on src/sklearn_evaluation/plot/style.py:
```diff
@@ -0,0 +1,165 @@
+    material_ui_colors = get_color_palette()
+    plt.rcParams["axes.prop_cycle"] = mpl.cycler(color=material_ui_colors)
+
+    if cmap_style == "gradient":
```
Should we print some warning message if the input `cmap_style` is not supported?

Like:
```
@apply_theme(cmap_style="odd_color")
def plot_bar_with_cmap(cmap=None):
...
```

It will print
```
Currently we only support `monochromatic` and `gradient` in cmap_style
```

tonykploomber commented on src/sklearn_evaluation/plot/classification.py:
```diff
@@ -30,9 +31,11 @@ def plot(self, ax=None):
+
         _plot_cm(
             self.cm,
-            cmap=default_heatmap(),
```
nit: I think we can keep this 
```
        _plot_cm(
            self.cm,
            cmap=default_heatmap(),
```

idomic replied:
@yafim please address

yafimvo replied:
We updated the default cmap with new colors, so using this function will return the old coloring. I kept this function since we might want to go back to the old coloring or toggle between them.

neelasha23 commented on docs/reference/developer.md:
```diff
@@ -242,6 +242,126 @@ def my_plotting_function(y_true, y_pred, ax=None):
 
+### Styling plots
+Using our context style feature we can change matplotlib `rcParams` and easily manipulate the overall visual style of our plots. The default style features a frame with borders only on the left and bottom, and a monochromatic color scheme of one color from the material UI color palette. However, we have the ability to adjust both the cmap and the style of the chart itself using the parameters 'cm ...
+
```
`only on the left and bottom **of the plot**`

neelasha23 commented on docs/reference/developer.md:
```diff
@@ -242,6 +242,126 @@ def my_plotting_function(y_true, y_pred, ax=None):
+
+    offset = 0
+    for i in range(len(data)):
+        color = mpl.colormaps.get_cmap(cmap)(float(i) / len(data))
```
We can also add example that passes `ax_style` as parameter to the decorator, and one where we pass both `ax_style` and `cmap_style`

edublancas commented:
i fixed the docs in master, please rebase

yafimvo commented:
@idomic fixed all rejects and rebased

edublancas commented on src/sklearn_evaluation/plot/feature_ranking.py:
```diff
@@ -197,6 +199,7 @@ def feature_ranks_custom_algorithm(self, ranks):
         return self.ax
 
 
+@apply_theme()
```
both this one and the 2D version are no longer displaying the docs correctly: https://sklearn-evaluation--279.org.readthedocs.build/en/279/api/plot.html#rank1d

perhaps it'll be fixed if we apply the decorator to RankD or if we apply it to the method that creates the plot?

edublancas reviewed (changes requested):
just a minor comment
